File: JavaScript/Algorithms/Stack Applications/balanced parentheses.js

let arePair = (c1, c2) => {
    if(c1 === '(' && c2 === ')')
        return true;
    else if(c1 === '[' && c2 === ']')
        return true;
    else if(c1 === '{' && c2 === '}')
        return true;
    else
        return false;
}

let isBalanced = (s) => {
    let stack = [];
    for(let i = 0; i < s.length; i++) {
        if(s[i] === '(' || s[i] === '[' || s[i] === '{') {
            stack.push(s[i]);
            continue;
        }

        if(s[i] === ')' || s[i] === ']' || s[i] === '}') {
            if(stack.length === 0 || !arePair(stack.pop(), s[i])) {
                return false;
            }
        }
    }
    return true;
}

console.log(isBalanced('{}'));
console.log(isBalanced('{()}'));
console.log(isBalanced('{)}'));
console.log(isBalanced('{(())()[()]}'));
console.log(isBalanced('{(())()[()])}'));
console.log(isBalanced('{(())()[(])}'));
console.log(isBalanced('{(())()[([{}])]}'));
console.log(isBalanced('}{(())()[()]}'));

File: C#/Algorithms/Stack Applications/balanced parentheses.cs
using System;
using System.Collections;
namespace Stack_Applications
{
    class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("{}: "+ isBalanced("{}"));
            Console.WriteLine("{()}: " + isBalanced("{()}"));
            Console.WriteLine("{)}: "+ isBalanced("{)}"));
            Console.WriteLine("{(}): " + isBalanced("{(})"));
            Console.WriteLine("{(())()[()]}: "+ isBalanced("{(())()[()]}"));
            Console.WriteLine("{(())()[([{}])]}: "+ isBalanced("{(())()[([{}])]}"));
            Console.WriteLine("((12+5)*4): "+ isBalanced("((12+5)*4)"));
            Console.WriteLine("((12+5)*(4)): "+ isBalanced("((12+5)*(4))"));
            Console.WriteLine("((12+5)*(4): "+ isBalanced("((12+5)*(4)"));
            Console.WriteLine("((12+5)*(4)/12): "+ isBalanced("((12+5)*(4)/12)"));
        }

        public static bool arePair(char c1, char c2)
        {
            if ((c1 == '(' && c2 == ')') || (c1 == '[' && c2 == ']') || (c1 == '{' && c2 == '}'))
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public static bool isBalanced(string s)
        {
            Stack stack = new Stack();
            int length = s.Length;
            for (int i = 0; i < length; i++)
            {
                char temp = s[i];
                if (temp == '(' || temp == '[' || temp == '{')
                {
                    stack.Push(temp);
                    continue;
                }

                if (temp == ')' || temp == ']' || temp == '}')
                {
                    if (stack.Count == 0 || !arePair(Convert.ToChar(stack.Pop()), temp))
                    {
                        return false;
                    }
                }
            }
            return stack.Count == 0;
        }
    }
}